src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetWebApi2Integration.cs

```diff
@@ -54,39 +56,49 @@ public static class AspNetWebApi2Integration
             if (apiController == null) { throw new ArgumentNullException(nameof(apiController)); }
 
             var cancellationToken = (CancellationToken)boxedCancellationToken;
-            return ExecuteAsyncInternal(apiController, controllerContext, cancellationToken, opCode, mdToken, moduleVersionPtr);
-        }
+            var callOpCode = (OpCodeValue)opCode;
+            var httpControllerType = apiController.GetInstrumentedInterface(HttpControllerTypeName);
 
-        /// <summary>
-        /// Calls the underlying ExecuteAsync and traces the request.
-        /// </summary>
-        /// <param name="apiController">The Api Controller</param>
-        /// <param name="controllerContext">The controller context for the call</param>
-        /// <param name="cancellationToken">The cancellation token</param>
-        /// <param name="opCode">The OpCode used in the original method call.</param>
-        /// <param name="mdToken">The mdToken of the original method call.</param>
-        /// <param name="moduleVersionPtr">A pointer to the module version GUID.</param>
-        /// <returns>A task with the result</returns>
-        private static async Task<HttpResponseMessage> ExecuteAsyncInternal(
-            object apiController,
-            object controllerContext,
-            CancellationToken cancellationToken,
-            int opCode,
-            int mdToken,
-            long moduleVersionPtr)
-        {
-            Func<object, object, CancellationToken, Task<HttpResponseMessage>> instrumentedMethod;
+            Type taskResultType;
 
             try
             {
-                var httpControllerType = apiController.GetInstrumentedInterface(HttpControllerTypeName);
+                var request = controllerContext.GetProperty<object>("Request").GetValueOrDefault();
+                var httpRequestMessageType = request.GetInstrumentedType("System.Net.Http.HttpRequestMessage");
 
-                instrumentedMethod = MethodBuilder<Func<object, object, CancellationToken, Task<HttpResponseMessage>>>
+                // The request should never be null, so get the base type found in System.Net.Http.dll
+                if (httpRequestMessageType != null)
+                {
+                    var systemNetHttpAssembly = httpRequestMessageType.Assembly;
+                    taskResultType = systemNetHttpAssembly.GetType("System.Net.Http.HttpResponseMessage", true);
+                }
+
+                // This should never happen, but put in a reasonable fallback of finding the first System.Net.Http.dll in the AppDomain
```

Comment:
Can this ever actually happen? Should we emit some sort of metric if it does?

Reply:
This could only happen if the `HttpControllerContext.Request` was null, but I don't think that's actually possible because in `System.Web.Http` there seems to be earlier null checks on that request object before constructing the owning `HttpControllerContext`. Sure, I can emit a metric here. What do you think of the following snippet:

```
                    var statsd = Tracer.Instance.Statsd;
                    if (statsd != null)
                    {
                        statsd.AppendWarning(source: $"{nameof(AspNetWebApi2Integration)}.{nameof(ExecuteAsync)}", message: "Using fallback logic to find System.Net.Http.HttpResponseMessage Type needed for return type", null);
                        statsd.Send();
                    }
```

Reply:
Done in d5625bf14ce1d0c13bac6283d76655e2e3c27d14